Implement Undo

nlowe opened the pull request:
Adds the ability to undo moves up to 10 turns ago as outlined in the project specification.
## TODO:
- [x] Verify undo of 10 levels (Initial inspection seems fine)
- [x] Fix animations (Regular movement works, (un)merging is weird)

Will close #6

Commit:
Initial work on game states

Commit:
Fix movement vectors for undoing existing cells

Commit:
Fix undo for newly merged cells.

techwiz24 commented:
Need to fix animations for un-merging then this PR can be merged.

techwiz24 commented:
Experiencing some inconsistencies with undoing. Sometimes an exception is thrown because the position stack has been exhausted.

Commit:
Fix edge case in undo

Some moves cannot be completely undone because the
position history buffer has been exhausted. Fix this
by making the buffer larger than the max undo size.

Also fixes stats being reset if all undo's are used.

techwiz24 commented:
Needs Re-Verified due to f04139a7c89704e93037c4720f79b10df687dcc3

Commit:
Fix undo animation for newly merged cells
